### PR TITLE
Pressure-velocity gradient decoupling (surface loss on pressure only)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The model predicts [Ux, Uy, p] jointly from a shared representation. But pressure and velocity have fundamentally different scales and physical meaning. Velocity errors on surface nodes are tiny (Ux/Uy near zero at no-slip surfaces) and add noise to the gradient without helping pressure. By detaching velocity channels from the surface loss, pressure gradients get an uncontaminated path through the network.

## Instructions
In the training loop, after computing `abs_err` (around line 623), change the surface loss computation to only backprop through the pressure channel.

Run: `python train.py --agent frieren --wandb_name "frieren/pressure-decouple" --wandb_group pressure-gradient-decouple`

## Baseline
- val/loss: 2.1997
- surf_p MAE: in_dist=20.03, ood_cond=20.57, tandem=40.41

---

## Results

Three iterations, same branch.

### Run 1: Full detachment (W&B run `9zpex4cm`, channel_weights=[0,0,1])
| Split | val/loss | surf_Ux | surf_Uy | surf_p |
|---|---|---|---|---|
| val_in_dist | 5.097 | 1.928 | 0.796 | **20.43** |
| val_ood_cond | 6.192 | 1.921 | 0.759 | **19.66** |
| val_ood_re | overflow | 1.490 | 0.630 | **30.37** |
| val_tandem | 8.332 | 2.855 | 1.272 | **42.17** |
| **combined val/loss** | **6.54** | | | |

Velocity +300-500%. No-slip learning destroyed.

### Run 2: Asymmetric [0.1, 0.1, 1.0] (W&B run `rgp8t8ks`)
| Split | val/loss | surf_Ux | surf_Uy | surf_p |
|---|---|---|---|---|
| val_in_dist | 1.842 | 0.419 | 0.233 | **21.05** |
| val_ood_cond | 2.245 | 0.390 | 0.264 | **19.05** |
| val_ood_re | overflow | 0.363 | 0.249 | **30.11** |
| val_tandem | 3.656 | 0.801 | 0.422 | **41.44** |
| **combined val/loss** | **2.5806** | | | |

Velocity +36%, pressure ood_cond -7.4%, ood_re -2.6%.

### Run 3: Asymmetric [0.5, 0.5, 1.5] (W&B run `rpfokxd1`)
| Split | val/loss | mae_surf_Ux | mae_surf_Uy | mae_surf_p | mae_vol_Ux | mae_vol_Uy | mae_vol_p |
|---|---|---|---|---|---|---|---|
| val_in_dist | 1.665 | 0.328 | 0.198 | **21.17** | 1.335 | 0.461 | 26.58 |
| val_ood_cond | 1.964 | 0.298 | 0.208 | **19.18** | 1.074 | 0.407 | 19.58 |
| val_ood_re | overflow | 0.304 | 0.213 | **30.11** | 1.059 | 0.442 | 51.03 |
| val_tandem_transfer | 3.441 | 0.696 | 0.369 | **41.79** | 2.199 | 1.014 | 43.59 |
| **combined val/loss** | **2.3569** | | | | | | |

**vs baseline (2.1997):**
| Metric | Baseline | Run 3 | Delta |
|---|---|---|---|
| val/loss | 2.1997 | 2.3569 | **+0.157 (+7.1%)** |
| surf_Ux in_dist | ~0.307 | 0.328 | **+0.021 (+7%)** |
| surf_p in_dist | 20.03 | 21.17 | **+1.14 (+5.7%)** |
| surf_p ood_cond | 20.57 | 19.18 | **-1.39 (-6.8%) ✓** |
| surf_p ood_re | 30.90 | 30.11 | **-0.79 (-2.6%) ✓** |
| surf_p tandem | 40.41 | 41.79 | **+1.38 (+3.4%)** |

**Epochs:** 66 | **Peak GPU memory:** ~10.6 GB

### Pattern across all three iterations

| Weights | val/loss | surf_Ux Δ | surf_p ood_cond | surf_p ood_re |
|---|---|---|---|---|
| Baseline [1,1,1] | 2.1997 | — | 20.57 | 30.90 |
| [0,0,1] (detach) | 6.54 | +523% | 19.66 | 30.37 |
| [0.1,0.1,1.0] | 2.5806 | +36% | 19.05 | 30.11 |
| [0.5,0.5,1.5] | 2.3569 | +7% | 19.18 | 30.11 |

**Consistent finding**: every iteration improves ood_cond pressure (-5 to -8%) and ood_re pressure (-2.6%). The cost is velocity surface accuracy and combined val/loss.

**Root tension**: The combined val/loss metric includes velocity surface accuracy. Any approach that shifts gradient budget from velocity to pressure will score worse on val/loss even when pressure improves. The "right" outcome for this task (engineers care about pressure) conflicts with the overall optimization target (val/loss).

With [0.5, 0.5, 1.5], the velocity degradation is small (+7%) while pressure ood_cond improves by -6.8% — but this still drives val/loss up +7.1%.

### Suggested follow-ups

- **Pressure-only surf_weight**: Compute separate surface pressure loss and surface velocity loss; use adaptive surf_weight only on pressure component. This cleanly separates the objective from velocity interference without changing the combined val/loss formulation.
- **Accept the tradeoff**: If the advisor's goal is purely surf_p reduction (not combined val/loss), the [0.5, 0.5, 1.5] weights with -6.8% ood_cond surf_p is already an improvement worth keeping despite the val/loss regression.

### CODE REVIEW

**Comment by tcapelle** [comment]
**Requesting changes — the pressure insight is real, execution too aggressive.**

The ood_cond surf_p improvement (-4.4%, 20.57→19.66) proves the core hypothesis: velocity noise in the surface loss gradient hurts pressure optimization. But fully detaching velocity killed velocity predictions (+500%).

**New approach: asymmetric per-channel surface weighting.** Instead of detaching velocity, downweight it 10x in the surface loss:

```python
# Asymmetric surface loss: 10x weight on pressure, 1x on velocity
channel_weights = torch.tensor([0.1, 0.1, 1.0], device=abs_err.device)  # [Ux, Uy, p]
abs_err_weighted = abs_err * channel_weights.unsqueeze(0).unsqueeze(0)
surf_per_sample = (abs_err_weighted * surf_mask.unsqueeze(-1)).sum(dim=(1, 2)) / surf_mask.sum(dim=1).clamp(min=1).float()
```

This preserves velocity learning on surface nodes (no-slip enforcement) while shifting 91% of the surface gradient budget to pressure. The model still sees that Ux≈Uy≈0 on surfaces, just with less emphasis.

Run: `python train.py --agent frieren --wandb_name "frieren/asymmetric-surf-10x" --wandb_group pressure-gradient-decouple`
